nautobot_netbox_importer/diffsync/nautobot.py

```diff
@@ -0,0 +1,362 @@
+        """Set default values for a Nautobot instance constructor."""
+        self.constructor_kwargs = defaults
+
+
+class ImporterModel(DiffSyncModel):
+    """Base class for all DiffSync models."""
+
+    _wrapper: NautobotModelWrapper
+
+    @classmethod
+    def create(cls, diffsync: BaseAdapter, ids: dict, attrs: dict) -> Optional[DiffSyncModel]:
+        """Create this model instance, both in Nautobot and in DiffSync."""
+        instance = cls._wrapper.model(**cls._wrapper.constructor_kwargs, **ids)
+        if not isinstance(diffsync, NautobotAdapter):
+            raise TypeError(f"Invalid diffsync type {diffsync}")
+        diffsync.save_nautobot_instance(instance, attrs, cls._wrapper.fields)
+        return super().create(diffsync, ids, attrs)
+
+    def update(self, attrs: dict) -> Optional[DiffSyncModel]:
+        """Update this model instance, both in Nautobot and in DiffSync."""
+        uid = getattr(self, "id", None)
+        if not uid:
+            raise NotImplementedError("Cannot update model without id")
+
+        model = self._wrapper.model
+        instance = model.objects.get(id=uid)
+        if not isinstance(self.diffsync, NautobotAdapter):
+            raise TypeError(f"Invalid diffsync type {self.diffsync}")
+        self.diffsync.save_nautobot_instance(instance, attrs, self._wrapper.fields)
+        return super().update(attrs)
+
+
+class NautobotAdapter(BaseAdapter):
+    """Nautobot DiffSync Adapter."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the adapter."""
+        super().__init__("Nautobot", *args, **kwargs)
+        self.clean_failures: Dict[NautobotBaseModelType, Set[Uid]] = {}
+        self.validation_errors: Optional[Dict[ContentTypeStr, Set[ValidationError]]] = None
+
+    def get_validation_errors(self) -> Dict[ContentTypeStr, Set[ValidationError]]:
+        """Re-run clean() on all instances that failed validation."""
+        if self.validation_errors is not None:
+            return self.validation_errors
+
+        self.validation_errors = {}
+
+        failures = self.clean_failures
+        self.clean_failures = {}
+
+        for model_type, uids in failures.items():
+            errors = None
+            for uid in uids:
+                instance = model_type.objects.get(id=uid)
+                try:
+                    instance.clean()
+                except ValidationError as error:
+                    if errors:
+                        errors.add(error)
+                    else:
+                        errors = set([error])
+                        self.validation_errors[model_type._meta.label.lower()] = errors  # type: ignore
+
+        return self.validation_errors
+
+    # pylint: disable=too-many-statements
+    def save_nautobot_instance(self, instance: NautobotBaseModel, values: RecordData, fields: NautobotFields) -> None:
+        """Save a Nautobot instance."""
+
+        def set_custom_field(value: Any):
```

Comment:
Looks like `value` is expected to be a Mapping, not Any.

Reply:
Fixed typing.